src/python/pants/backend/shell/lint/shellcheck/subsystem.py

```diff
@@ -43,6 +44,16 @@ def register_options(cls, register):
                 "Arguments to pass directly to Shellcheck, e.g. `--shellcheck-args='-e SC20529'`.'"
             ),
         )
+        register(
+            "--config",
+            type=file_option,
+            advanced=True,
+            help=(
+                "Path to `.shellcheckrc` file.\n\nBecause Shellcheck does not have a config file "
+                "option, you must locate this file somewhere Shellcheck can auto-discover it, "
```

Comment:
I still don't understand. No matter where the user puts the file under the build root, we can still ignore that relative location and always materialize it in the process chroot root.

Reply:
Agreed. But if they aren't careful to put it in the correct location, shfmt's auto discovery will fail and they may be confused. This has confused people in the past with isort, given that other tools like flake8 allow you to put the config anywhere and Pants will set a --config option for the tool for it to all work.

In fact, this is precisely what motivated Cristian with the isort PR.

Reply:
OK - wow. I still do not understand. Sorry for being dense. My understaning is that this flag points to a pathc to a shellcheckrc file. Technically, as I understand it, that file could be anwhere under the build root (only due to pants current limitations therin), and it could even have any name. All that matters is it has the right content. We then slurp that up and plop it in `<process chroot>/.shellcheckrc`. Now that leaves the fact that shellcheck won't work as expected when run outside of pants since `.shellcheckrc` is not in the right place or doesn't have the right name. That though is not our problem, right?

Reply:
> Technically, as I understand it, that file could be anwhere under the build root (only due to pants current limitations therin), and it could even have any name. All that matters is it has the right content. We then slurp that up and plop it in <process chroot>/.shellcheckrc. 

Aha! Yes, that is indeed correct. We could do that. 

But if we _want_ to do that is another question. The precedent was set with isort to not, as the location does change things. For example, our old `contrib/.isort.cfg` only applied to `contrib/**` it would be very wrong for Pants to do something magical. Also with shfmt, Pytest, and Shellcheck, the location does matter and you can set in subdirectories so that only part of your code uses the config. 

So, doing the relocation would make a tradeoff to allow people to consolidate config files in one place, but break the workflow where you specifically placed the config in a subdirectory so that only some files use it. Perhaps the solution is to offer both via some mechanism like a second option or string prefix - but no users have asked for that relocation yet, so I'd bias towards keeping this.

Here, I want to ensure our help message is clear that the location of the config file does indeed matter.

Reply:
Ok. So, by forcing the user to specify one config file we're actually breaking their setup if they use nested. So since we know about this nested thing in .isort.cfg, .shellcheckrc, etc it seems a little bit cruel to not just do the right thing and kill the option. We can look for the closest nested config file, and if one exists, copy it into the chroot at that location. Or, if you prefer a chroot sandbox that is probably less confusing to reason about as a user, just copy all the nested config files found back up the ancestor path.

Reply:
For isort, we instead use a list option so that you can specify all of them. 

Atm, Pants never auto-detects and includes config files, although that warning PR does get us closer to that direction. I'm skeptical we want to start doing that for two reasons:

1) for other tools like Flake8 with arbitrary config locations, we can never auto-discover perfectly. We will always need a config file option for those. For those ones like Flake8, would we attempt to auto-discover and the option is only for cases we can't figure out? If so, how do you disable? I don't like the inconsistency of why the option is sometimes used and sometimes not. 
2) how do you turn off this feature? A Boolean option is possibly too coarse - how do you opt out of only one or two? Perhaps I'm being overly cautious and a Boolean is fine. 

I'm not super concerned about the clunkiness of a user having to tell Pants about which config files to use, particularly because it's usually only a one time thing on initial setup. It'd also declarative. I _am_ concerned about people not realizing they need to tell Pants, hence the new warning.

What do you think? Make this a list option or pursue auto discovery?

Reply:
Always auto-discover when its unambiguous. This is by definition a per-tool thing. You need to straighten out your priorities I think. Is Pants just runs tools trumping over Pants does things consistently? I think so, and as such Pants should let you use tools like you already use them whenever possible and not force you to add boilerplate.

Reply:
As a user, that would make me most happy. That way I need never learn anything about Pants - configuration, etc - unless I really have to. And in all those cases, it should logically be clear why - "Oh right, frooboz tool lets me do random things - there is no way Pants could have guessed that - it makes sense I need to configure this knob".

Reply:
Fair enough on it Just Working. This is another instance where we can reduce the cost to adopt Pants, similar to dep inference and tailor.

What do you think about how you turn off this behavior for tools without a config file option? A Boolean option should be sufficient I think?

For tools like Flake8 and MyPy, do we start to auto-include if we can detect it? I think yes, let's lean into this. Which leaves the question of how to model disabling the auto-detection while still allowing for a manual config file - a single option --config can't cover both things. 

Perhaps for all tools, we use --config-discovery, which is a bool. Some also have --config for a manual value.

Reply:
I'm a bit lost on "how you turn it off". Is the you a programmer or a user? Presumably you never turn it off as a user - you just delete the config file! As a programmer this is just an API question - you're good at that and can figure something out.

> For tools like Flake8 and MyPy, do we start to auto-include if we can detect it? 

If the accepted trumping principle is don't get in the way of how a tool works on its own / don't force boilerplate - I think the question answeres itself.